Modules/_sqlite/connection.c

```diff
@@ -440,40 +442,56 @@ remove_callbacks(sqlite3 *db)
 static void
 connection_close(pysqlite_Connection *self)
 {
-    if (self->db) {
-        if (self->autocommit == AUTOCOMMIT_DISABLED &&
-            !sqlite3_get_autocommit(self->db))
-        {
-            /* If close is implicitly called as a result of interpreter
-             * tear-down, we must not call back into Python. */
-            if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
-                remove_callbacks(self->db);
-            }
-            (void)connection_exec_stmt(self, "ROLLBACK");
+    assert(self->db);
+    if (self->autocommit == AUTOCOMMIT_DISABLED &&
+        !sqlite3_get_autocommit(self->db))
+    {
+        /* If close is implicitly called as a result of interpreter
+         * tear-down, we must not call back into Python. */
+        if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
+            remove_callbacks(self->db);
         }
+        (void)connection_exec_stmt(self, "ROLLBACK");
+    }
 
-        free_callback_contexts(self);
+    free_callback_contexts(self);
 
-        sqlite3 *db = self->db;
-        self->db = NULL;
+    sqlite3 *db = self->db;
+    self->db = NULL;
 
-        Py_BEGIN_ALLOW_THREADS
-        int rc = sqlite3_close_v2(db);
-        assert(rc == SQLITE_OK), (void)rc;
-        Py_END_ALLOW_THREADS
+    Py_BEGIN_ALLOW_THREADS
+    int rc = sqlite3_close_v2(db);
+    assert(rc == SQLITE_OK), (void)rc;
+    Py_END_ALLOW_THREADS
+}
+
+static void
+connection_finalize(PyObject *self)
+{
+    pysqlite_Connection *con = (pysqlite_Connection *)self;
+    PyObject *exc = PyErr_GetRaisedException();
+    /* Clean up if user has not called .close() explicitly. */
+    if (con->db) {
+        if (PyErr_ResourceWarning(self, 1, "unclosed %R", self)) {
+            /* Spurious errors can appear at shutdown */
+            if (PyErr_ExceptionMatches(PyExc_Warning)) {
```

Comment:
Why only log Warning exception? Is it a good thing to call connection_close() with an exception set?

Reply:
I followed the pattern from other places in the code base which uses this approach :)

Reply:
https://github.com/python/cpython/blob/2da67a5789a9247ae6611185bf786b697aa04fbe/Modules/posixmodule.c#L15234-L15256

and

https://github.com/python/cpython/blob/2da67a5789a9247ae6611185bf786b697aa04fbe/Modules/socketmodule.c#L5223-L5253

Reply:
Why not simply call `PyErr_WriteUnraisable` if `PyErr_ResourceWarning` fails? I don't get it.

Reply:
Maybe these two files should be fixed to remove ExceptionMatches() call?

Reply:
I suspect there is a reason for those guards; I'd like to find out that first.

Reply:
They were added in respectively ffe96ae10be8a3117fa18c35034fcfc45c3cf7b7 and e033e06db077d5abcb4bc3729d03f8a4a09b2486 by Antoine and Serhiy.

Reply:
Let's handle this post merge.